
# Laboratorio 4(Aplicacion de Regularizacion - Regresion Logistica one-vs-all) Grupo 1
 <h3>En este laboratorio se hizo el uso del dataset para aplicar la regularización y sin utilizar regularización a la Regresion Logistica one-vs-all, y como siguiente se presento los resultados de ambas experiencias<h3>
 <HR>
 <h3>
  NOMBRE: POLO ORELLANA BRAYAN SIMON <br>
  CARRERA: INGENIERIA DE SISTEMAS <BR>
  FECHA: 02/04/2024 <BR>

  * [Enlace de invitacion para ser colaborador](https://github.com/bspoloo/SIS420-012024/invitations)
  
  * [Enlace al git hub](https://github.com/bspoloo/SIS420-012024/tree/main/Laboratorios/Laboratorio%204)
  
  * [Enlace al Colab](https://colab.research.google.com/github/bspoloo/SIS420-012024/blob/main/Laboratorios/Laboratorio%204/Regresion_logistica_one_vs_all.ipynb?hl=es)
 <h3>
 

Para el laboratorio se hizo uso del para entrenar el modelo de **Regresion Logistica one vs all** aplicando regularizacion y dada la información relacionada con el crédito de una persona, se creo un modelo de aprendizaje automático que pueda clasificar el puntaje crediticio de una persona

 A lo largo de los años, se han recopilado datos bancarios básicos y mucha información relacionada con el crédito de varias personas. Se quiere construir un sistema inteligente para segregar a las personas en tramos de puntuación crediticia para reducir los esfuerzos manuales. Haciendo uso del siguiente dataset: [credit_score_classification_processed](https://www.kaggle.com/datasets/deepaksaipendyala/credit-score-classification-processed).

El archivo `credit_score.csv` contiene un conjunto de datos de entrenamiento para clasificar el puntaje crediticio de una persona.

Primero se importo todas las librerias necesarias:

In [316]:
# utilizando la libreria os para manejos de directorios
import os

# Computacion vectorial y cientifica para python
import numpy as np  

#importamos pandas para el manejo del dataset, y separarlos dentro de una matriz
import pandas as pd

#esta tabulate nos sirve para hacer tablas
from tabulate import tabulate

# Librerias para graficación (trazado de gráficos)
from matplotlib import pyplot
from mpl_toolkits.mplot3d import Axes3D  # -> Necesario para graficar superficies 3D

#Para separa el 20% y 80%
from sklearn.model_selection import train_test_split

# Modulo de optimizacion en scipy
from scipy import optimize

## Carga de datos del dataset

cargamos los datos haciendo el uso de la libreria **Pandas** que  es una herramienta poderosa y versátil utilizada para manipulación y análisis de datos. Ofrece estructuras de datos flexibles y eficientes para trabajar con datos tabulares, como hojas de cálculo en Excel o tablas SQL. Algunas de las funcionalidades clave de pandas incluyen:

In [317]:
#Haciendo uso de la libreria pandas para leer el dataset, delimitado por ","
df = pd.read_csv('credit_score.csv', delimiter=',')
df

Unnamed: 0  Month        Age  Occupation  Annual_Income  \
0               1      2  23.000000          13       19114.12   
1               2      6  34.429817          13       19114.12   
2               3      0  23.000000          13       19114.12   
3               4      7  23.000000          13       19114.12   
4               5      5  23.000000          13       19114.12   
...           ...    ...        ...         ...            ...   
79800       99991      1  29.000000           1       20002.88   
79801       99993      2  25.000000           9       39628.99   
79802       99997      5  25.000000           9       39628.99   
79803       99998      4  25.000000           9       39628.99   
79804       99999      1  25.000000           9       39628.99   

       Monthly_Inhand_Salary  Num_Bank_Accounts  Num_Credit_Card  \
0                4194.170850                3.0              4.0   
1                4194.170850                3.0              4.0   
2                4194.170850                3.0              4.0   
3                1824.843333                3.0              4.0   
4                4194.170850                3.0              4.0   
...                      ...                ...              ...   
79800            1929.906667               10.0              8.0   
79801            3359.415833                4.0              6.0   
79802            3359.415833                4.0              6.0   
79803            3359.415833                4.0              6.0   
79804            3359.415833                4.0              6.0   

       Interest_Rate  Num_of_Loan  ...  Credit_Mix  Outstanding_Debt  \
0           3.000000          4.0  ...           1            809.98   
1           3.000000          4.0  ...           1            809.98   
2           3.000000          4.0  ...           1            809.98   
3           3.000000          4.0  ...           1            809.98   
4           3.000000          4.0  ...           1            809.98   
...              ...          ...  ...         ...               ...   
79800      29.000000          5.0  ...           0           3571.70   
79801       7.000000          2.0  ...           1            502.38   
79802      14.546679          2.0  ...           1            502.38   
79803       7.000000          2.0  ...           1            502.38   
79804       7.000000          2.0  ...           1            502.38   

       Credit_Utilization_Ratio  Credit_History_Age  Payment_of_Min_Amount  \
0                     31.944960                   0                      1   
1                     28.609352                 267                      1   
2                     31.377862                 268                      1   
3                     24.797347                 269                      1   
4                     27.262259                 270                      1   
...                         ...                 ...                    ...   
79800                 37.140784                  75                      2   
79801                 29.135447                 376                      1   
79802                 41.255522                 380                      1   
79803                 33.638208                 381                      1   
79804                 34.192463                 382                      1   

       Total_EMI_per_month  Amount_invested_monthly  Payment_Behaviour  \
0                49.574949               118.280222                  3   
1                49.574949                81.699521                  4   
2                49.574949               199.458074                  5   
3                49.574949                41.420153                  1   
4                49.574949                62.430172                  6   
...                    ...                      ...                ...   
79800            60.964772                34.662906                  0   
798

In [318]:
#eliminamos la primera columan que seria nuestro "id" ya que no nos sirve para el analisis, que esa columna es unnamed
df = df.drop(df.columns[0], axis=1)

#ahora tenemos un dataframe
#Imprimimos en una tabla el dataset para hacer un analisis mas claro.
df

Month        Age  Occupation  Annual_Income  Monthly_Inhand_Salary  \
0          2  23.000000          13       19114.12            4194.170850   
1          6  34.429817          13       19114.12            4194.170850   
2          0  23.000000          13       19114.12            4194.170850   
3          7  23.000000          13       19114.12            1824.843333   
4          5  23.000000          13       19114.12            4194.170850   
...      ...        ...         ...            ...                    ...   
79800      1  29.000000           1       20002.88            1929.906667   
79801      2  25.000000           9       39628.99            3359.415833   
79802      5  25.000000           9       39628.99            3359.415833   
79803      4  25.000000           9       39628.99            3359.415833   
79804      1  25.000000           9       39628.99            3359.415833   

       Num_Bank_Accounts  Num_Credit_Card  Interest_Rate  Num_of_Loan  \
0                    3.0              4.0       3.000000          4.0   
1                    3.0              4.0       3.000000          4.0   
2                    3.0              4.0       3.000000          4.0   
3                    3.0              4.0       3.000000          4.0   
4                    3.0              4.0       3.000000          4.0   
...                  ...              ...            ...          ...   
79800               10.0              8.0      29.000000          5.0   
79801                4.0              6.0       7.000000          2.0   
79802                4.0              6.0      14.546679          2.0   
79803                4.0              6.0       7.000000          2.0   
79804                4.0              6.0       7.000000          2.0   

       Type_of_Loan  ...  Credit_Mix  Outstanding_Debt  \
0               128  ...           1            809.98   
1               128  ...           1            809.98   
2               128  ...           1            809.98   
3               128  ...           1            809.98   
4               128  ...           1            809.98   
...             ...  ...         ...               ...   
79800          4913  ...           0           3571.70   
79801           683  ...           1            502.38   
79802           683  ...           1            502.38   
79803           683  ...           1            502.38   
79804           683  ...           1            502.38   

       Credit_Utilization_Ratio  Credit_History_Age  Payment_of_Min_Amount  \
0                     31.944960                   0                      1   
1                     28.609352                 267                      1   
2                     31.377862                 268                      1   
3                     24.797347                 269                      1   
4                     27.262259                 270                      1   
...                         ...                 ...                    ...   
79800                 37.140784                  75                      2   
79801                 29.135447                 376                      1   
79802                 41.255522                 380                      1   
79803                 33.638208                 381                      1   
79804                 34.192463                 382                      1   

       Total_EMI_per_month  Amount_invested_monthly  Payment_Behaviour  \
0                49.574949               118.280222                  3   
1                49.574949                81.699521                  4   
2                49.574949               199.458074                  5   
3                49.574949                41.420153                  1   
4                49.574949                62.430172                  6   
...                    ...                      ...                ...   
79800            60.964772                34.662906  

## Analisis del dataset
Hacemos un analisis del dataset mostrando su informacion usando la funcion de `info()`:

In [319]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79805 entries, 0 to 79804
Data columns (total 24 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Month                     79805 non-null  int64  
 1   Age                       79805 non-null  float64
 2   Occupation                79805 non-null  int64  
 3   Annual_Income             79805 non-null  float64
 4   Monthly_Inhand_Salary     79805 non-null  float64
 5   Num_Bank_Accounts         79805 non-null  float64
 6   Num_Credit_Card           79805 non-null  float64
 7   Interest_Rate             79805 non-null  float64
 8   Num_of_Loan               79805 non-null  float64
 9   Type_of_Loan              79805 non-null  int64  
 10  Delay_from_due_date       79805 non-null  int64  
 11  Num_of_Delayed_Payment    79805 non-null  float64
 12  Changed_Credit_Limit      79805 non-null  int64  
 13  Num_Credit_Inquiries      79805 non-null  float64
 14  Credit

mostrando la tabla nos damos cuenta que nuestra Y a predicir es el **Credit_Score**, para tener una mejor vision de la cantidad de clases que existe, se hizo el siguiente codigo:

donde separamos nuestra columna Y con `value_counts()` este método de pandas cuenta el número de veces que aparece cada valor único en la columna **Credit_Score** del DataFrame. Devuelve una Serie pandas donde los índices son los valores únicos de la columna **Credit_Score** y los valores son el recuento de ocurrencias de cada valor.

In [320]:
#para contar cuantas clases contiene nuestra columna y
class_counts = df["Credit_Score"].value_counts()

#para contar cuantas caracteristicas tiene nuestro dataset, obviamente con sin contar nuestra y, por eso lo dropeamos, tambien dropeamos la primera columna que no tiene nombre
feactures_counts =df.drop(['Credit_Score'], axis=1)
feactures_counts = feactures_counts.shape[1]

print(f"la cantidad de caracteristicas es: { feactures_counts}")
print(f'la cantidad de clases es: 3')

#mostramos la cantidad de clases tiene, y en que cantidad
class_counts

la cantidad de caracteristicas es: 23
la cantidad de clases es: 3


Credit_Score
1    42470
0    23129
2    14206
Name: count, dtype: int64

Hay 3 clases diferentes de puntajes crediticios: etiquetados como 0, 1 y 2.

* La clase 1 parece ser la más común, seguida por la clase 0 y luego la clase 2.
* El recuento de cada clase es el siguiente:
    - Clase 0: 23129 personas.
    - Clase 1: 42470 personas.
    - Clase 2: 14206 personas.
* Este análisis de la distribución de los puntajes crediticios en el conjunto de datos es útil para comprender la proporción de cada clase. Esto puede ser relevante al desarrollar un modelo de clasificación como la Regresión Logística one vs all. 

In [321]:
#almacenamos la cantidad de caracteristicas en una variable
input_layer_size  = feactures_counts;

#almacenamos la cantidad de clases en una variable

#para este caso no es necesario cambiar los valores de la ultima clase, ya que este cuenta con 0,1,2, 
# en caso de contener 1,2,3 se tendria que hacer los cambios necesarios para que sea 1,2,0, o simplememte suma 1 a la cantidad de clases.
num_labels = 3;

## Separacion del 80% de los datos para entrenamiento y 20% para pruebas

Se debe tomar en cuenta que cada clase tiene su propia cantidad, por lo cual separarlos directamente en un 80% para entrenamiento y un 20% para test no seria tan efectivo, ya que puede que en el 80% hay mas datos de una clase que las otras, provocando que nuestro modelo no conozca mucho sobre esa clase, por lo cual debe separarse un 80% para entrenamiento y un 20% para pruebas de cada clase.

Para este hacemos uso de la libreria **train_test_split**

In [322]:
# hacemos uso del DataFrame llamado 'df' que contiene nuestros datos datos
# y es la columna que contiene las etiquetas de clase (en este caso, 'Credit_Score')

#creamos una variable temporal que contentra toda la columna de 'Credit_Score'
y_temp = df['Credit_Score']

# Para la clase 0

#donde y_temp es igual a 0, separamos los datos en train_class_0, test_class_0
data_class_0 = df[y_temp == 0]
train_class_0, test_class_0 = train_test_split(data_class_0, test_size=0.2, random_state=42)

# Para la clase 1

#donde y_temp es igual a 1, separamos los datos en train_class_1, test_class_1
data_class_1 = df[y_temp == 1]
train_class_1, test_class_1 = train_test_split(data_class_1, test_size=0.2, random_state=42)

# Para la clase 2

#donde y_temp es igual a 2, separamos los datos en train_class_2, test_class_2
data_class_2 = df[y_temp == 2]
train_class_2, test_class_2 = train_test_split(data_class_2, test_size=0.2, random_state=42)

haciendo conteo de separacion de datos:

In [323]:
print(f"Para la clase 0 tenemos una cantidad de: { data_class_0.shape[0]} donde el 80% es: {train_class_0.shape[0]} y el 20% es: {test_class_0.shape[0]}")
print(f"Para la clase 1 tenemos una cantidad de: { data_class_1.shape[0]} donde el 80% es: {train_class_1.shape[0]} y el 20% es: {test_class_1.shape[0]}")
print(f"Para la clase 2 tenemos una cantidad de: { data_class_2.shape[0]} donde el 80% es: {train_class_2.shape[0]} y el 20% es: {test_class_2.shape[0]}")
print(f"La cantidad total de datos es: {df.shape[0]}")

Para la clase 0 tenemos una cantidad de: 23129 donde el 80% es: 18503 y el 20% es: 4626
Para la clase 1 tenemos una cantidad de: 42470 donde el 80% es: 33976 y el 20% es: 8494
Para la clase 2 tenemos una cantidad de: 14206 donde el 80% es: 11364 y el 20% es: 2842
La cantidad total de datos es: 79805


Separamos cada uno en sus X_train, y_train, X_test y y_test respectivos:

In [324]:
#para la parte de entrenamiento, separamos las caracteristicas de la etiqueta

#para la clase 0
X_train_class_0 = train_class_0.drop(['Credit_Score'], axis=1)
y_train_class_0 = train_class_0['Credit_Score']

#para la clase 1
X_train_class_1 = train_class_1.drop(['Credit_Score'], axis=1)
y_train_class_1 = train_class_1['Credit_Score']

#para la clase 2
X_train_class_2 = train_class_2.drop(['Credit_Score'], axis=1)
y_train_class_2 = train_class_2['Credit_Score']

#ahora para la parte de pruebas, separamos las caracteristicas de la etiqueta
#para la clase 0
X_test_class_0 = test_class_0.drop(['Credit_Score'], axis=1)
y_test_class_0 = test_class_0['Credit_Score']

#para la clase 1
X_test_class_1 = test_class_1.drop(['Credit_Score'], axis=1)
y_test_class_1 = test_class_1['Credit_Score']

#para la clase 2
X_test_class_2 = test_class_2.drop(['Credit_Score'], axis=1)
y_test_class_2 = test_class_2['Credit_Score']

ahora unimos todos en una sola matriz para X y y:
pero luego debemos mezclar los datos, haciendo uso de ``np.random.permutation(len(X))`` genera un arreglo de índices permutados aleatoriamente.
Luego, estos índices se usan para reorganizar tanto las características como las etiquetas de entrenamiento y prueba.

In [325]:
#separando los datos de entrenamiento y pruebas

#para los datos de entrenamiento
X_train = pd.concat([X_train_class_0, X_train_class_1, X_train_class_2]).values
y_train = pd.concat([y_train_class_0, y_train_class_1, y_train_class_2]).values


indices_train = np.random.permutation(len(X_train))
X_train = X_train[indices_train]
y_train = y_train[indices_train]
m_train = len(y_train)

#para los datos de pruebas
X_test = pd.concat([X_test_class_0, X_test_class_1, X_test_class_2]).values
y_test = pd.concat([y_test_class_0, y_test_class_1, y_test_class_2]).values

indices_test = np.random.permutation(len(X_test))
X_test = X_test[indices_test]
y_test = y_test[indices_test]
m_test = len(y_test)


Visualizamos algunos datos de entrenamiento:

In [326]:
#Visualizamos una cantidad de datos de entrenamiento
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>6s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]', 'X[:, 10]', 'X[:, 11]', 'X[:, 12]', 
    'X[:, 13]', 'X[:, 14]', 'X[:, 15]', 'X[:, 16]', 'X[:, 17]', 'X[:, 18]', 'X[:, 19]', 'X[:, 20]', 'X[:, 21]', 'X[:, 22]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}'.format(
        X_train[i, 0],
        X_train[i, 1],
        X_train[i, 2], 
        X_train[i, 3], 
        X_train[i, 4], 
        X_train[i, 5], 
        X_train[i, 6], 
        X_train[i, 7], 
        X_train[i, 8], 
        X_train[i, 9], 
        X_train[i, 10],
        X_train[i, 11],
        X_train[i, 12], 
        X_train[i, 13], 
        X_train[i, 14], 
        X_train[i, 15], 
        X_train[i, 16],
        X_train[i, 17],
        X_train[i, 18],
        X_train[i, 19], 
        X_train[i, 20], 
        X_train[i, 21], 
        X_train[i, 22], 
        y_train[i]
    ))

print(" ")
print('El 80% de ejemplos para entrenamiento son la cantidad de: {:.0f} de ejemplos'.format( len(X_train)))
print('El 20% de ejemplos para pruebas son la cantidad de: {:.0f} de ejemplos'.format( len(X_test)))
print('La cantidad total de ejemplos es de: {:.0f} de ejemplos'.format( len(df)))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]  X[:, 10]  X[:, 11]  X[:, 12]  X[:, 13]  X[:, 14]  X[:, 15]  X[:, 16]  X[:, 17]  X[:, 18]  X[:, 19]  X[:, 20]  X[:, 21]  X[:, 22]     Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
         3        25        12     15280      4194         3       7        15         4      5255        28         9      4015         3         2       563        34       236         2        51       130         5       211       1
         3        34        12     65984      5279         8       3        24         4      5379        15        19      1286        12         2      1377        39       107         2       170       637         2       383       0
         3        40        10     206

## Funcion para la Normalización de caracteristicas

Al visualizar los datos se puede observar que las caracteristicas tienen diferentes magnitudes, por lo cual se debe transformar cada valor en una escala de valores similares, esto con el fin de que el descenso por el gradiente pueda converger mas rapidamente. En este caso

Hacemos el uso de la siguiente funcion para normalizar los datos:

In [327]:
def  featureNormalize(X):
    X_norm = X.copy()

    #creamos un array de ceros con una longitud igual al número de columnas en el array X. La variable mu y sigma se inicializa como este array de ceros.
    mu = np.zeros(X.shape[1])
    sigma = np.zeros(X.shape[1])

    #Creamos el promedio de cada filaa de X
    #media de cada columna
    mu = np.mean(X, axis = 0)
    
    #desviacion estandar de cada fila de X
    sigma = np.std(X, axis = 0)
    
    sigma[sigma == 0] = 1
    
    #normalizamos los datos con la siguiente formula
    X_norm = (X - mu) / sigma

    return X_norm, mu, sigma

## Funcion para el calculo de la sigmoide

También conocida como la función logística, es una función matemática que toma cualquier número real como entrada y devuelve un valor en el rango de 0 a 1. Donde nuestra **Z** es nuestra hipotesis.

In [328]:
def sigmoid(z):
    z = np.array(z)

    g = np.zeros(z.shape)

    g = 1 / (1 + np.exp(-z))

    return g

## Funcion de calculo de costo con regularizacion

Aplicando la teoria donde nuestra funcion recibira parametros como Theta, x, y y lamda_, donde lamda_ es nuestro parametro de regularizacion.

Donde la funcion nos devuelve un costo y nuestro gradiente.

In [329]:
def lrCostFunctionCR(theta, X, y, lambda_):
    #creamos una variable m que contiene la longitud de y
    m = y.size

    # convierte las etiquetas a valores enteros si son boleanos
    if y.dtype == bool:
        y = y.astype(int)

    #inicializamos J y grad
    J = 0
    grad = np.zeros(theta.shape)

    #calculamos h haciendo uso de la funcion sigmoid, donde h es nuestra hipotesis
    h = sigmoid(X.dot(theta.T))

    temp = theta
    temp[0] = 0

    J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))

    grad = (1 / m) * (h - y).dot(X)
    
    grad = grad + (lambda_ / m) * temp

    return J, grad

### Funcion de One-vs-all con regularizacion

In [330]:
def oneVsAllCR(X, y, num_labels, lambda_):
    # algunas variables utiles
    #m es la longitud de y, n es la cantidad de columnas en X
    m, n = X.shape

    # inicializamos la matriz de thetas con ceros, y con n+1 columnas que seria nuestro sesgo
    all_theta = np.zeros((num_labels, n + 1))

    # Agrega unos a la matriz X
    X = np.concatenate([np.ones((m, 1)), X], axis=1)

    # iteramos sobre cada etiqueta (clase) y entrenamos un clasificador
    for c in np.arange(num_labels):
        initial_theta = np.zeros(n + 1)
        options = {'maxiter': 20000}
        res = optimize.minimize(lrCostFunctionCR,
                                initial_theta,
                                (X, (y == c), lambda_),
                                jac=True,
                                method='CG',
                                options=options)

        all_theta[c] = res.x

    return all_theta

## Funcion de calculo de costo sin regularizacio

Para este caso solo quitamos la sumatoria `+ (lambda_ / (2 * m)) * np.sum(np.square(temp))` a la ecuacion de costo, donde esto se aplicado sin regularizar

In [331]:
def lrCostFunctionSR(theta, X, y):
    #creamos una variable m que contiene la longitud de y
    m = y.size

    # convierte las etiquetas a valores enteros si son boleanos
    if y.dtype == bool:
        y = y.astype(int)

    #inicializamos J y grad
    J = 0
    
    # grad = np.zeros(theta.shape)

    #calculamos h haciendo uso de la funcion sigmoid, donde h es nuestra hipotesis
    h = sigmoid(X.dot(theta.T))

    # temp = theta
    # temp[0] = 0

    # J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))
    J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h)))

    grad = (1 / m) * (h - y).dot(X)
    
    # grad = grad + (lambda_ / m) * temp
    
    return J, grad

### Funcion de One-vs-all sin regularizacion

Como esta funcion sera sin regularizar, ya no se necesita usar el parametro `lambda_` por lo tanto al momento de obtener las thetas para cada clase ya no es necesario enviarle un parametro `lambda_`:

In [332]:
def oneVsAllSR(X, y, num_labels):
    # algunas variables utiles
    #m es la longitud de y, n es la cantidad de columnas en X
    m, n = X.shape

    # inicializamos la matriz de thetas con ceros, y con n+1 columnas que seria nuestro sesgo
    all_theta = np.zeros((num_labels, n + 1))

    # Agrega unos a la matriz X
    X = np.concatenate([np.ones((m, 1)), X], axis=1)

    # iteramos sobre cada etiqueta (clase) y entrenamos un clasificador
    for c in np.arange(num_labels):
        initial_theta = np.zeros(n + 1)
        options = {'maxiter': 20000}
        res = optimize.minimize(lrCostFunctionSR,
                                initial_theta,
                                (X, (y == c)),
                                jac=True,
                                method='CG',
                                options=options)

        all_theta[c] = res.x

    return all_theta

## Funcion para Prediccion One-vs-all
Aqui creamos solo la funcion, donde mandamos los parametros de `all_theta` y la `X` que en este caso puede ser las X de prueba, pero antes deben de estar normalizadas.

In [333]:
def predictOneVsAll(all_theta, X):
    m = X.shape[0];
    num_labels = all_theta.shape[0]

    # inicializamos la matriz de p con ceros
    p = np.zeros(m)

    # añadimos unos a la matriz de X
    X = np.concatenate([np.ones((m, 1)), X], axis=1)
    p = np.argmax(sigmoid(X.dot(all_theta.T)), axis = 1)

    return p

## 1. Usando regularizacion

La regularización es una técnica utilizada en el aprendizaje automático para prevenir el **sobreajuste (overfitting)** de un modelo a los datos de entrenamiento.

El sobreajuste ocurre cuando un modelo se ajusta demasiado bien a los datos de entrenamiento y captura el ruido o las fluctuaciones aleatorias en los datos en lugar de aprender la verdadera relación subyacente entre las características y la variable objetivo. Esto puede resultar en un rendimiento deficiente del modelo cuando se enfrenta a nuevos datos que no formaban parte del conjunto de entrenamiento.

### 1.1 Clasificacion Multiclase

Carga de los datos para la clasificacion multiclase:

In [334]:
#estos datos seran usados para el entrenamiento
X_testCR = X_test.copy()
y_testCR = y_test.copy()
m_test_CR = len(y_testCR)

#estos datos seran usados para el test
X_trainCR = X_train.copy()
y_trainCR = y_train.copy()
m_train_CR = len(y_trainCR)

#Visualizamos una cantidad de datos de entrenamiento
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>6s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]', 'X[:, 10]', 'X[:, 11]', 'X[:, 12]', 
    'X[:, 13]', 'X[:, 14]', 'X[:, 15]', 'X[:, 16]', 'X[:, 17]', 'X[:, 18]', 'X[:, 19]', 'X[:, 20]', 'X[:, 21]', 'X[:, 22]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}'.format(
        X_trainCR[i, 0],
        X_trainCR[i, 1],
        X_trainCR[i, 2], 
        X_trainCR[i, 3], 
        X_trainCR[i, 4], 
        X_trainCR[i, 5], 
        X_trainCR[i, 6], 
        X_trainCR[i, 7], 
        X_trainCR[i, 8], 
        X_trainCR[i, 9], 
        X_trainCR[i, 10],
        X_trainCR[i, 11],
        X_trainCR[i, 12], 
        X_trainCR[i, 13], 
        X_trainCR[i, 14], 
        X_trainCR[i, 15], 
        X_trainCR[i, 16],
        X_trainCR[i, 17],
        X_trainCR[i, 18],
        X_trainCR[i, 19], 
        X_trainCR[i, 20], 
        X_trainCR[i, 21], 
        X_trainCR[i, 22], 
        y_trainCR[i]
    ))

print(" ")
print('El 80% de ejemplos para entrenamiento son la cantidad de: {:.0f} de ejemplos'.format( len(X_train)))
print('El 20% de ejemplos para pruebas son la cantidad de: {:.0f} de ejemplos'.format( len(X_test)))
print('La cantidad total de ejemplos es de: {:.0f} de ejemplos'.format( len(df)))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]  X[:, 10]  X[:, 11]  X[:, 12]  X[:, 13]  X[:, 14]  X[:, 15]  X[:, 16]  X[:, 17]  X[:, 18]  X[:, 19]  X[:, 20]  X[:, 21]  X[:, 22]     Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
         3        25        12     15280      4194         3       7        15         4      5255        28         9      4015         3         2       563        34       236         2        51       130         5       211       1
         3        34        12     65984      5279         8       3        24         4      5379        15        19      1286        12         2      1377        39       107         2       170       637         2       383       0
         3        40        10     206

#### 1.1.1 Normalizacion de las caracteristicas

Se hace uso de la funcion de `featureNormalize(X) ` donde se recibe un parametro de tipo matriz para normalizar cada dato dentro de ella, retornandome la **matriz normalizda**, **sigma(desviacion estandar)**, y mi **mu(media)**.

Almacenando los datos normalizados en **X_norm** usando la funcion **featureNormaliza()**, normalizando los datos de X_train

In [335]:
X_norm_CR, mu_CR, sigma_CR = featureNormalize(X_trainCR)

# imprimir todos las X_norm de datos solo 10
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>6s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]', 'X[:, 10]', 'X[:, 11]', 'X[:, 12]', 
    'X[:, 13]', 'X[:, 14]', 'X[:, 15]', 'X[:, 16]', 'X[:, 17]', 'X[:, 18]', 'X[:, 19]', 'X[:, 20]', 'X[:, 21]', 'X[:, 22]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:8.0f}'.format(
        X_norm_CR[i, 0],
        X_norm_CR[i, 1],
        X_norm_CR[i, 2], 
        X_norm_CR[i, 3], 
        X_norm_CR[i, 4], 
        X_norm_CR[i, 5], 
        X_norm_CR[i, 6], 
        X_norm_CR[i, 7], 
        X_norm_CR[i, 8], 
        X_norm_CR[i, 9], 
        X_norm_CR[i, 10],
        X_norm_CR[i, 11],
        X_norm_CR[i, 12], 
        X_norm_CR[i, 13], 
        X_norm_CR[i, 14], 
        X_norm_CR[i, 15], 
        X_norm_CR[i, 16],
        X_norm_CR[i, 17],
        X_norm_CR[i, 18],
        X_norm_CR[i, 19], 
        X_norm_CR[i, 20], 
        X_norm_CR[i, 21], 
        X_norm_CR[i, 22], 
        y_trainCR[i]
    ))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]  X[:, 10]  X[:, 11]  X[:, 12]  X[:, 13]  X[:, 14]  X[:, 15]  X[:, 16]  X[:, 17]  X[:, 18]  X[:, 19]  X[:, 20]  X[:, 21]  X[:, 22]     Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     -0.21     -0.97      0.98     -0.11     -0.00     -0.81      0.47      0.05      0.18      0.89      0.47     -0.10      1.21     -0.13      0.97     -0.75      0.33      0.31      0.86     -0.16     -0.26      1.01     -0.90       1
     -0.21      0.00      0.98     -0.08      0.37      0.88     -0.89      1.09      0.18      0.95     -0.41     -0.05     -1.00     -0.08      0.97     -0.05      1.31     -0.82      0.86     -0.15     -0.00     -0.51     -0.10       0
     -0.21      0.58      0.55    

### 1.2 Clasificacion One-vs-All
Aqui estariamos usando un ciclo for para iterar sobre cada una de las clases, luego haciendo uso de `optimize.minimize` que es un método de la biblioteca *scipy* que encuentra el mínimo de una función. En este caso, se trata de minimizar la función de costos de regresión logística `(lrCostFunction)`.

Los parámetros iniciales ``(initial_theta)``.
Una tupla que contiene los datos de entrenamiento ``(X)``, las etiquetas ``(y == c)``, y el parámetro de regularización ``(lambda_)``.
``jac=True`` indica que la función de coste devuelve tanto el coste como el gradiente.
El método de optimización ``('CG' significa Gradiente Conjugado).``
El diccionario de opciones ``(options)``  que se establece en 1000 para limitar el número máximo de iteraciones del optimizador.


Inicializamos nuestra lambda con valor de *0.005*, usamos la funcion de `oneVsAll` donde pasamos los parametros de *X_norm*, *num_labels* que seria la cantidad de clases que tenemos, y nuestro *lanbda* para asi obtener nuestros Thetas para cada clase.

In [336]:

lambda_ = 1000
all_theta_CR = oneVsAllCR(X_norm_CR, y_trainCR, num_labels, lambda_)
print(all_theta_CR.shape)

#imprimimos todos los thetas para cada clase
for i in range(all_theta_CR.shape[1]):
    print("Thetha",i," para la clase 0:", all_theta_CR[0,i]," ; ", "Thetha",i," para la clase 1:", all_theta_CR[1,i]," ; ", "Thetha",i," para la clase 2:", all_theta_CR[2,i])

(3, 24)
Thetha 0  para la clase 0: -1.0695169567537348  ;  Thetha 0  para la clase 1: 0.14044304105950767  ;  Thetha 0  para la clase 2: -2.186541760724309
Thetha 1  para la clase 0: 0.001099483554564459  ;  Thetha 1  para la clase 1: 0.015678220045970335  ;  Thetha 1  para la clase 2: -0.031174327875785328
Thetha 2  para la clase 0: -0.05732951636146707  ;  Thetha 2  para la clase 1: 0.004432269652256636  ;  Thetha 2  para la clase 2: 0.053675683652515446
Thetha 3  para la clase 0: 0.006515575044222821  ;  Thetha 3  para la clase 1: -0.0006646074117552987  ;  Thetha 3  para la clase 2: -0.012853200853254738
Thetha 4  para la clase 0: -0.022314452167680125  ;  Thetha 4  para la clase 1: 0.018915736516325072  ;  Thetha 4  para la clase 2: -0.0007569880100460452
Thetha 5  para la clase 0: -0.07393637295873941  ;  Thetha 5  para la clase 1: -0.00569475992649457  ;  Thetha 5  para la clase 2: 0.09343492800043075
Thetha 6  para la clase 0: 0.0057267758499581585  ;  Thetha 6  para la clase 1

### 1.3 Validaciones
Para hacer las validaciones correspondientes se hizo el uso siguiendo el consejo de 80/20, donde 80% es para la fase de entrenamiento, y 20% es para la fase de prueba.

Para este caso de validacion se hizo uso de la funcion de `predictOneVsAll()` creada anteriormente, los siguiente fue normalizar nuestra `X_test` que son el 20% para pruebas, ahora se procedio a normalizarlo haciendo uso de *mu* y *sigma* calculado anteriormente en la funcion de normalizacion.

luego se hace uso de `np.mean` donde nos calcula el promedio de los valores booleanos en el array resultante de la comparación. 

#### 1.3.1 Con los datos de entrenamiento
Para este caso se uso los datos de `X_norm` usando la funcion de `predictOneVsAll()`, viendo el resultado de precision nos da un **61.20483%** el cual esta muy lejos de ser un buen modelo.

In [337]:
# print(X_test.shape)
pred_test_CR = predictOneVsAll(all_theta_CR, X_norm_CR)
print('Precision del conjunto de entrenamiento: {:.5f}%'.format(np.mean(pred_test_CR == y_trainCR) * 100))

Precision del conjunto de entrenamiento: 61.20483%


#### 1.3.2 con los datos de prueba
Para este caso se uso los datos de `X_test`, luego se procedio a normalizar los datos. 

Usando la funcion de `predictOneVsAll()`, viendo el resultado de precision nos da un **61.41461%** , un poco en diferencia a la anterior prediccion.

In [338]:
X_test1_CR = X_testCR.copy()
# print(X_test.shape)
X_test1_CR = (X_test1_CR - mu_CR) / sigma_CR
pred_train_CR = predictOneVsAll(all_theta_CR, X_test1_CR)
print('Precision del conjunto de entrenamiento: {:.5f}%'.format(np.mean(pred_train_CR == y_testCR) * 100))

Precision del conjunto de entrenamiento: 61.41461%


En este caso se puede hacer las predicciones usando directamente la funcion `sigmoid()`, y luego imprimimos 30 predicciones para ver que tanto se asemejan con nuestra Y predicha.

In [339]:

#usando el mu y sigma calculado anteriormente, para realizar la normalizacion de los datos de prueba
X_test1_CR = (X_testCR.copy() - mu_CR) / sigma_CR
X_test1_CR = np.concatenate([np.ones((len(X_test1_CR), 1)), X_test1_CR], axis=1)

print(X_test1_CR.shape)

#para este ejemplo, vamos a mostrar las predicciones de las primeras 30 predicciones y compararlas con las reales
p = np.argmax(sigmoid(X_test1_CR.dot(all_theta_CR.T)), axis = 1)

for i in range(30):
    print(f"Prediccion: {p[i]}, Real: {y_test[i]}")

(15962, 24)
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 0
Prediccion: 0, Real: 0
Prediccion: 1, Real: 1
Prediccion: 2, Real: 2
Prediccion: 2, Real: 2
Prediccion: 0, Real: 0
Prediccion: 1, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 1, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 2, Real: 0
Prediccion: 2, Real: 2
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 2, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 0
Prediccion: 1, Real: 2
Prediccion: 0, Real: 1


## 2. Usando sin regularizacion

En esta parte del laboratorio no se usara la parte de regularizacion, esto para ver como es el comportamiento del modelo.

### 2.1 Clasificacion Multiclase

Carga de los datos para la clasificacion multiclase sin usar regularizacion:

In [340]:
#estos datos seran usados para el entrenamiento
X_testSR = X_test.copy()
y_testSR = y_test.copy()
m_test_SR = len(y_testSR)

#estos datos seran usados para el test
X_trainSR = X_train.copy()
y_trainSR = y_train.copy()
m_train_SR = len(y_trainSR)

#Visualizamos una cantidad de datos de entrenamiento
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>6s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]', 'X[:, 10]', 'X[:, 11]', 'X[:, 12]', 
    'X[:, 13]', 'X[:, 14]', 'X[:, 15]', 'X[:, 16]', 'X[:, 17]', 'X[:, 18]', 'X[:, 19]', 'X[:, 20]', 'X[:, 21]', 'X[:, 22]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:10.0f}{:8.0f}'.format(
        X_trainSR[i, 0],
        X_trainSR[i, 1],
        X_trainSR[i, 2], 
        X_trainSR[i, 3], 
        X_trainSR[i, 4], 
        X_trainSR[i, 5], 
        X_trainSR[i, 6], 
        X_trainSR[i, 7], 
        X_trainSR[i, 8], 
        X_trainSR[i, 9], 
        X_trainSR[i, 10],
        X_trainSR[i, 11],
        X_trainSR[i, 12], 
        X_trainSR[i, 13], 
        X_trainSR[i, 14], 
        X_trainSR[i, 15], 
        X_trainSR[i, 16],
        X_trainSR[i, 17],
        X_trainSR[i, 18],
        X_trainSR[i, 19], 
        X_trainSR[i, 20], 
        X_trainSR[i, 21], 
        X_trainSR[i, 22], 
        y_trainSR[i]
    ))

print(" ")
print('El 80% de ejemplos para entrenamiento son la cantidad de: {:.0f} de ejemplos'.format( len(X_train)))
print('El 20% de ejemplos para pruebas son la cantidad de: {:.0f} de ejemplos'.format( len(X_test)))
print('La cantidad total de ejemplos es de: {:.0f} de ejemplos'.format( len(df)))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]  X[:, 10]  X[:, 11]  X[:, 12]  X[:, 13]  X[:, 14]  X[:, 15]  X[:, 16]  X[:, 17]  X[:, 18]  X[:, 19]  X[:, 20]  X[:, 21]  X[:, 22]     Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
         3        25        12     15280      4194         3       7        15         4      5255        28         9      4015         3         2       563        34       236         2        51       130         5       211       1
         3        34        12     65984      5279         8       3        24         4      5379        15        19      1286        12         2      1377        39       107         2       170       637         2       383       0
         3        40        10     206

#### 2.1.1 Normalizacion de las caracteristicas

Se hace uso de la funcion de `featureNormalize(X) ` donde se recibe un parametro de tipo matriz para normalizar cada dato dentro de ella, retornandome la **matriz normalizda**, **sigma(desviacion estandar)**, y mi **mu(media)**.

Almacenando los datos normalizados en **X_norm** usando la funcion **featureNormaliza()**, normalizando los datos de X_train

In [341]:
X_norm_SR, mu_SR, sigma_SR = featureNormalize(X_trainSR)

# imprimir todos las X_norm de datos solo 10
print('{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>10s}{:>6s}'.format(
    'X[:,0]', 'X[:, 1]', 'X[:, 2]', 'X[:, 3]', 'X[:, 4]', 'X[:, 5]', 'X[:, 6]', 'X[:, 7]', 'X[:, 8]', 'X[:, 9]', 'X[:, 10]', 'X[:, 11]', 'X[:, 12]', 
    'X[:, 13]', 'X[:, 14]', 'X[:, 15]', 'X[:, 16]', 'X[:, 17]', 'X[:, 18]', 'X[:, 19]', 'X[:, 20]', 'X[:, 21]', 'X[:, 22]','Y'
))
print('-' * 250)

for i in range(10):
    print('{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:10.2f}{:8.0f}'.format(
        X_norm_SR[i, 0],
        X_norm_SR[i, 1],
        X_norm_SR[i, 2], 
        X_norm_SR[i, 3], 
        X_norm_SR[i, 4], 
        X_norm_SR[i, 5], 
        X_norm_SR[i, 6], 
        X_norm_SR[i, 7], 
        X_norm_SR[i, 8], 
        X_norm_SR[i, 9], 
        X_norm_SR[i, 10],
        X_norm_SR[i, 11],
        X_norm_SR[i, 12], 
        X_norm_SR[i, 13], 
        X_norm_SR[i, 14], 
        X_norm_SR[i, 15], 
        X_norm_SR[i, 16],
        X_norm_SR[i, 17],
        X_norm_SR[i, 18],
        X_norm_SR[i, 19], 
        X_norm_SR[i, 20], 
        X_norm_SR[i, 21], 
        X_norm_SR[i, 22], 
        y_trainSR[i]
    ))

    X[:,0]   X[:, 1]   X[:, 2]   X[:, 3]   X[:, 4]   X[:, 5]   X[:, 6]   X[:, 7]   X[:, 8]   X[:, 9]  X[:, 10]  X[:, 11]  X[:, 12]  X[:, 13]  X[:, 14]  X[:, 15]  X[:, 16]  X[:, 17]  X[:, 18]  X[:, 19]  X[:, 20]  X[:, 21]  X[:, 22]     Y
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
     -0.21     -0.97      0.98     -0.11     -0.00     -0.81      0.47      0.05      0.18      0.89      0.47     -0.10      1.21     -0.13      0.97     -0.75      0.33      0.31      0.86     -0.16     -0.26      1.01     -0.90       1
     -0.21      0.00      0.98     -0.08      0.37      0.88     -0.89      1.09      0.18      0.95     -0.41     -0.05     -1.00     -0.08      0.97     -0.05      1.31     -0.82      0.86     -0.15     -0.00     -0.51     -0.10       0
     -0.21      0.58      0.55    

### 2.2 Clasificacion One-vs-All

In [342]:
all_theta_SR = oneVsAllSR(X_norm_SR, y_trainSR, num_labels)
print(all_theta_CR.shape)

#imprimimos todos los thetas para cada clase
for i in range(all_theta_SR.shape[1]):
    print("Thetha",i," para la clase 0:", all_theta_SR[0,i]," ; ", "Thetha",i," para la clase 1:", all_theta_SR[1,i]," ; ", "Thetha",i," para la clase 2:", all_theta_SR[2,i])

(3, 24)
Thetha 0  para la clase 0: -1.0815432135459961  ;  Thetha 0  para la clase 1: 0.14285349851581414  ;  Thetha 0  para la clase 2: -2.3711422465984513
Thetha 1  para la clase 0: 0.0014111368182945564  ;  Thetha 1  para la clase 1: 0.016860057646807064  ;  Thetha 1  para la clase 2: -0.03736459358204417
Thetha 2  para la clase 0: -0.06071942258364846  ;  Thetha 2  para la clase 1: 0.0071704012203908695  ;  Thetha 2  para la clase 2: 0.049803981495049006
Thetha 3  para la clase 0: 0.007170505160669617  ;  Thetha 3  para la clase 1: -0.0011672879403555906  ;  Thetha 3  para la clase 2: -0.01486555887904792
Thetha 4  para la clase 0: -0.024927809372523757  ;  Thetha 4  para la clase 1: 0.02115291750067745  ;  Thetha 4  para la clase 2: -0.0010770765438633222
Thetha 5  para la clase 0: -0.09457664390799475  ;  Thetha 5  para la clase 1: -0.004293653625063567  ;  Thetha 5  para la clase 2: 0.12319884596775409
Thetha 6  para la clase 0: -0.02054703063516854  ;  Thetha 6  para la clase 1

### 2.3 Validaciones
Para hacer las validaciones correspondientes se hizo el uso siguiendo el consejo de 80/20, donde 80% es para la fase de entrenamiento, y 20% es para la fase de prueba.

Para este caso de validacion se hizo uso de la funcion de `predictOneVsAll()` creada anteriormente, los siguiente fue normalizar nuestra `X_test` que son el 20% para pruebas, ahora se procedio a normalizarlo haciendo uso de *mu* y *sigma* calculado anteriormente en la funcion de normalizacion.

luego se hace uso de `np.mean` donde nos calcula el promedio de los valores booleanos en el array resultante de la comparación. 

#### 2.3.1 Con los datos de entrenamiento
Para este caso se uso los datos de `X_norm` usando la funcion de `predictOneVsAll()`, viendo el resultado de precision nos da un **61.99897%** el cual esta muy lejos de ser un buen modelo.

In [343]:
# print(X_test.shape)
pred_test_SR = predictOneVsAll(all_theta_SR, X_norm_SR)
print('Precision del conjunto de entrenamiento: {:.5f}%'.format(np.mean(pred_test_SR == y_trainSR) * 100))

Precision del conjunto de entrenamiento: 61.99897%


#### 1.3.2 con los datos de prueba
Para este caso se uso los datos de `X_test`, luego se procedio a normalizar los datos. 

Usando la funcion de `predictOneVsAll()`, viendo el resultado de precision nos da un **62.23531%** , un poco en diferencia a la anterior prediccion.

In [344]:
X_test1_SR = X_testSR.copy()
# print(X_test.shape)
X_test1_SR = (X_test1_SR - mu_SR) / sigma_SR
pred_train_SR = predictOneVsAll(all_theta_SR, X_test1_SR)
print('Precision del conjunto de entrenamiento: {:.5f}%'.format(np.mean(pred_train_SR == y_testSR) * 100))

Precision del conjunto de entrenamiento: 62.23531%


En este caso se puede hacer las predicciones usando directamente la funcion `sigmoid()`, y luego imprimimos 30 predicciones para ver que tanto se asemejan con nuestra Y predicha.

In [345]:
#usando el mu y sigma calculado anteriormente, para realizar la normalizacion de los datos de prueba
X_test1_SR = (X_testSR.copy() - mu_SR) / sigma_SR
X_test1_SR = np.concatenate([np.ones((len(X_test1_SR), 1)), X_test1_SR], axis=1)

print(X_test1_SR.shape)

#para este ejemplo, vamos a mostrar las predicciones de las primeras 30 predicciones y compararlas con las reales
p = np.argmax(sigmoid(X_test1_SR.dot(all_theta_SR.T)), axis = 1)

for i in range(30):
    print(f"Prediccion: {p[i]}, Real: {y_test[i]}")

(15962, 24)
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 0
Prediccion: 0, Real: 0
Prediccion: 1, Real: 1
Prediccion: 2, Real: 2
Prediccion: 2, Real: 2
Prediccion: 0, Real: 0
Prediccion: 1, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 0
Prediccion: 1, Real: 1
Prediccion: 0, Real: 0
Prediccion: 0, Real: 1
Prediccion: 0, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 2, Real: 0
Prediccion: 2, Real: 2
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 2, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 1
Prediccion: 1, Real: 0
Prediccion: 1, Real: 2
Prediccion: 0, Real: 1


## Conclusiones

En esta parte se presentaran las conclusiones de acuerdo a las experiencias vistas en este laboratorio aplicando regularizacion y sin aplicar regularizacion.

### ¿Cual es el mejor?

Se comparó el desempeño de modelos de regresión logistica one-vs-all con y sin regularización, se pueden extraer varias conclusiones significativas:

##### 1. Impacto de la regularización en la precisión del modelo:

 Se observa que la regularización puede tener un efecto en la precisión del modelo. En algunos casos, la regularización puede mejorar la precisión del modelo al reducir el sobreajuste (overfitting),
 especialmente cuando se tienen conjuntos de datos pequeños o altamente ruidosos. 
 
 Sin embargo, en conjuntos de datos más grandes o con menos ruido, la regularización puede no proporcionar mejoras significativas en la precisión y podría incluso degradarla.

 Para este caso se hizo la comparacion entre la preciscion de ambos modelos, con y sin regularizacion:

| Tipo  | Precision del modelo con regularizacion| Presicion del modelo sin regularizacion| Dieferencia entre ambos|
|---------|------|---------|-----------|
| Con los datos de entrenamiento |61.20483%  | 61.99897% | -0.79414% |
| Con los datos de prueba | 61.41461%  | 62.23531% | -0.8207% |

Sepodemos observar que tanto la precisión del modelo con regularización como la del modelo sin regularización son bastante similares en ambos conjuntos de datos, tanto en los datos de entrenamiento como en los de prueba. Sin embargo, hay una ligera diferencia entre ambos en cada conjunto de datos.

##### 2. Tiempo de entrenamiento

En algunos casos, la regularización puede aumentar el tiempo de entrenamiento del modelo debido a la necesidad de calcular los términos adicionales de penalización en la función de costo y actualizar los coeficientes del modelo de manera iterativa.

Haciendo una tabla para ver los cambios:

| Modelo  | Tiempo de entrenamiento del modelo con regularizacion | Tiempo de entrenamiento del modelo sin regularizacion |Dieferencia entre ambos|
|---------|------|---------|-----------|
| Clasificacion Multiclase |0.2 seg  | 0.3 seg | -0.1 seg|

Como se puede observar el tiempo de entrenamiento para el modelo usando regularizacion es con muy poco menor al usado sin regularizacion, con una diferencia del 0.1 segundos. Esto sugiere que la regularización puede haber contribuido a una reducción en el tiempo de entrenamiento del modelo. Esto puede ser atribuido a la capacidad de la regularización para simplificar el modelo y reducir la complejidad, lo que a su vez puede conducir a un proceso de entrenamiento más rápido.

#### ¿Entonces cual es el mejor?

Una vez realizadas las comparaciones del desempeño de modelos con y sin regularización en términos de tiempo de entrenamiento y precisión, se puede determinar que el mejor modelo es aquel que aplica regularización.

**¿Por qué?**

El tiempo de entrenamiento es menor en el modelo con regularización, a pesar de que su precisión es ligeramente inferior a la del modelo sin aplicar regularización. En muchos casos, especialmente cuando se trabaja con grandes conjuntos de datos, el tiempo de entrenamiento puede ser un factor crítico a considerar, ya que puede afectar significativamente la eficiencia del proceso de entrenamiento.

### ¿Cuales son los mejores hiperparametros?

Se eligio aquellos hiperparametros de acuerdo al tiempo y presicion que nos daba el modelo:

- **Numero de iteraciones.-** se eligio como una cantidad de **20000** iteraciones, esto guiandonos con la precision y el tiempo que tardaba en entrenar el modelo, si se elegia un numero de iteraciones arriba de 20000 el tiempo de entrenamiento era muy alto, y la presicion incluso era demasiado alejado del 100%.

- **Lambda.-** Para el hiperparametro **lambda** se uso el valor de **1000**, igualmente se siguio las recomendaciones que el hiperparametro lambda podria estar entre 10, 100 o 1000, se hizo la prueba con cada uno pero el que mejor resultados pero con mayor tiempo de entrenamiento y procesamiento nos dio fue el valor de un lamdda de *1000*. Se hizo la prueba con el valor de 10 lo cual nos daba un precision cercana a 62.24%, luego se probo con el valor de 100 el cual nos daba una precision de 61.998% y finalmente se uso el valor de 1000 y este nos dio una presicion del 61.20483%, el cual se selecciono el valor de 1000 para lambda, esto porque se busca un parametro un poco lejano a 0. Tambien se hizo la prueba con un valor de lambda cercano a cero, y nos genero una presicion igual al de la presicion del modelo sin regularizar, mientras que si usabamos un valor lambda negativo nos generaba una grafica contraria al de la curva esperada.

### ¿Cual asumirias en una determinada circunstancia?

Además de la mínima diferencia en los tiempos de entrenamiento y la precisión entre ambos modelos, la regularización proporciona una ventaja adicional al ayudar a prevenir el sobreajuste o overfitting, especialmente cuando se incrementa el número de iteraciones o se trabaja con conjuntos de datos más complejos. La regularización actúa como una medida preventiva para garantizar que el modelo generalice bien a datos no vistos y evite ajustarse demasiado a los datos de entrenamiento.

Dado que el tiempo de entrenamiento es corto para ambos modelos y el impacto en los recursos computacionales es mínimo, aplicar regularización es una elección sensata que puede mejorar la capacidad del modelo para generalizar a nuevos datos sin comprometer significativamente el rendimiento del entrenamiento.